File: backend/models/media.py
from sqlalchemy import Column, Integer, String, Boolean, Float, Text, ForeignKey, JSON, DateTime
from sqlalchemy.orm import relationship
from datetime import datetime
from backend.core.db import Base

class Library(Base):
    __tablename__ = "libraries"
    
    id = Column(Integer, primary_key=True, index=True)
    name = Column(String, index=True)
    path = Column(String, unique=True, index=True)
    type = Column(String) # 'movie' or 'tv'
    language = Column(String, default='en')

class Movie(Base):
    __tablename__ = "movies"

    id = Column(Integer, primary_key=True, index=True)
    library_id = Column(Integer, ForeignKey("libraries.id"))
    title = Column(String, index=True)
    sort_title = Column(String)
    original_title = Column(String)
    year = Column(Integer, index=True)
    
    # IDs
    tmdb_id = Column(String, index=True, nullable=True)
    imdb_id = Column(String, index=True, nullable=True)
    
    # Metadata
    plot = Column(Text, nullable=True)
    tagline = Column(String, nullable=True)
    genres = Column(JSON, nullable=True)
    cast = Column(JSON, nullable=True)
    director = Column(String, nullable=True)
    content_rating = Column(String, nullable=True) # e.g. PG-13, R
    runtime = Column(Integer, nullable=True) # minutes
    
    # Ratings
    tmdb_rating = Column(Float, nullable=True)
    imdb_rating = Column(Float, nullable=True)
    imdb_votes = Column(Integer, nullable=True)
    metascore = Column(Integer, nullable=True)
    
    # Images
    poster_path = Column(String, nullable=True)
    fanart_path = Column(String, nullable=True)
    
    # Status
    status = Column(String, default="unmatched") # unmatched, matched
    nfo_generated = Column(Boolean, default=False)
    file_renamed = Column(Boolean, default=False)
    trailer_url = Column(String, nullable=True)  # YouTube URL
    
    # Relationships
    files = relationship("MovieFile", back_populates="movie", cascade="all, delete-orphan")

class MovieFile(Base):
    __tablename__ = "movie_files"
    
    id = Column(Integer, primary_key=True, index=True)
    movie_id = Column(Integer, ForeignKey("movies.id"))
    file_path = Column(String, unique=True, index=True)
    original_filename = Column(String)
    size_bytes = Column(Integer)
    
    # Media Info Specs
    resolution = Column(String, nullable=True) # 1080p, 4K
    hdr_type = Column(String, nullable=True) # HDR10, DV
    video_codec = Column(String, nullable=True) # x264, x265
    audio_codec = Column(String, nullable=True) # TrueHD, DTS
    audio_channels = Column(String, nullable=True) # 5.1, 7.1
    part_number = Column(Integer, nullable=True) # For cd1, cd2
    subtitle_path = Column(String, nullable=True)  # Path to downloaded .srt
    
    movie = relationship("Movie", back_populates="files")

class TVShow(Base):
    __tablename__ = "tv_shows"
    
    id = Column(Integer, primary_key=True, index=True)
    library_id = Column(Integer, ForeignKey("libraries.id"))
    title = Column(String, index=True)
    year = Column(Integer, index=True)
    
    # IDs
    tmdb_id = Column(String, index=True, nullable=True)
    tvdb_id = Column(String, index=True, nullable=True)
    imdb_id = Column(String, index=True, nullable=True)
    
    # Metadata
    plot = Column(Text, nullable=True)
    genres = Column(JSON, nullable=True)
    cast = Column(JSON, nullable=True)
    director = Column(String, nullable=True)
    runtime = Column(Integer, nullable=True)
    content_rating = Column(String, nullable=True)
    imdb_rating = Column(Float, nullable=True)
    
    # Images
    poster_path = Column(String, nullable=True)
    fanart_path = Column(String, nullable=True)
    
    status = Column(String, default="unmatched")
    episode_ordering = Column(String, default="aired") # aired, dvd, absolute
    trailer_url = Column(String, nullable=True)
    
    seasons = relationship("Season", back_populates="show", cascade="all, delete-orphan")

class Season(Base):
    __tablename__ = "seasons"
    
    id = Column(Integer, primary_key=True, index=True)
    show_id = Column(Integer, ForeignKey("tv_shows.id"))
    season_number = Column(Integer, index=True)
    episode_count = Column(Integer, default=0)
    poster_path = Column(String, nullable=True)
    
    show = relationship("TVShow", back_populates="seasons")
    episodes = relationship("Episode", back_populates="season", cascade="all, delete-orphan")

class Episode(Base):
    __tablename__ = "episodes"
    
    id = Column(Integer, primary_key=True, index=True)
    season_id = Column(Integer, ForeignKey("seasons.id"))
    episode_number = Column(Integer, index=True)
    title = Column(String)
    plot = Column(Text, nullable=True)
    air_date = Column(String, nullable=True)
    
    file_path = Column(String, nullable=True, unique=True)
    original_filename = Column(String, nullable=True)
    thumbnail_path = Column(String, nullable=True)
    
    # Media Info Specs
    resolution = Column(String, nullable=True)
    video_codec = Column(String, nullable=True)
    audio_codec = Column(String, nullable=True)
    
    missing = Column(Boolean, default=False)
    
    season = relationship("Season", back_populates="episodes")

class BackgroundTask(Base):
    __tablename__ = "background_tasks"
    
    id = Column(String, primary_key=True, index=True)
    name = Column(String)
    status = Column(String, default="queued") # queued, running, done, error
    progress = Column(Integer, default=0)
    total = Column(Integer, default=100)
    message = Column(String, nullable=True)
    duration = Column(Float, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

File: internal/models/media.go
package models

import (
	"database/sql/driver"
	"encoding/json"
	"errors"
	"time"
)

type StringList []string

func (sl *StringList) Scan(value interface{}) error {
	bytes, ok := value.([]byte)
	if !ok {
		return errors.New("type assertion to []byte failed")
	}
	return json.Unmarshal(bytes, sl)
}

func (sl StringList) Value() (driver.Value, error) {
	return json.Marshal(sl)
}

type CastMember struct {
	Name  string `json:"name"`
	Role  string `json:"role"`
	Thumb string `json:"thumb"`
}

type Library struct {
	ID       uint   `gorm:"primaryKey;index" json:"id"`
	Name     string `gorm:"index" json:"name"`
	Path     string `gorm:"uniqueIndex" json:"path"`
	Type     string `json:"type"`
	Language string `gorm:"default:'en'" json:"language"`
}

type Movie struct {
	ID             uint       `gorm:"primaryKey;index" json:"id"`
	LibraryID      uint       `gorm:"index" json:"library_id"`
	Library        Library    `json:"library"`
	Title          string     `gorm:"index" json:"title"`
	SortTitle      string     `json:"sort_title"`
	OriginalTitle  string     `json:"original_title"`
	Year           int        `gorm:"index" json:"year"`
	TmdbID         string     `gorm:"index" json:"tmdb_id"`
	ImdbID         string     `gorm:"index" json:"imdb_id"`
	Plot           string     `json:"plot"`
	Tagline        string     `json:"tagline"`
	Genres         StringList `gorm:"type:text" json:"genres"`
	Cast           StringList `gorm:"type:text" json:"cast"`
	CastDetails    []CastMember `gorm:"serializer:json" json:"cast_details"`
	Director       string     `json:"director"`
	ContentRating  string     `json:"content_rating"`
	Runtime        int        `json:"runtime"`
	TmdbRating     float64    `json:"tmdb_rating"`
	ImdbRating     float64    `json:"imdb_rating"`
	TmdbVotes      int        `json:"tmdb_votes"`
	ImdbVotes      int        `json:"imdb_votes"`
	Metascore      int        `json:"metascore"`
	PosterPath     string     `json:"poster_path"`
	FanartPath     string     `json:"fanart_path"`
	Status         string     `gorm:"default:'unmatched'" json:"status"`
	NfoGenerated   bool       `gorm:"default:false" json:"nfo_generated"`
	FileRenamed    bool       `gorm:"default:false" json:"file_renamed"`
	TrailerURL     string     `json:"trailer_url"`
	Files          []MovieFile `gorm:"foreignKey:MovieID" json:"files"`
}

type MovieFile struct {
	ID               uint   `gorm:"primaryKey;index" json:"id"`
	MovieID          uint   `gorm:"index" json:"movie_id"`
	FilePath         string `gorm:"uniqueIndex" json:"file_path"`
	OriginalFilename string `json:"original_filename"`
	SizeBytes        int64  `json:"size_bytes"`
	Resolution       string `json:"resolution"`
	HdrType          string `json:"hdr_type"`
	VideoCodec       string `json:"video_codec"`
	AudioCodec       string `json:"audio_codec"`
	AudioChannels    string `json:"audio_channels"`
	PartNumber       int    `json:"part_number"`
	SubtitlePath     string `json:"subtitle_path"`
}

type TVShow struct {
	ID              uint       `gorm:"primaryKey;index" json:"id"`
	LibraryID       uint       `gorm:"index" json:"library_id"`
	Library         Library    `json:"library"`
	Title           string     `gorm:"index" json:"title"`
	Year            int        `gorm:"index" json:"year"`
	TmdbID          string     `json:"tmdb_id"`
	TvdbID          string     `json:"tvdb_id"`
	ImdbID          string     `json:"imdb_id"`
	Plot            string     `json:"plot"`
	Genres          StringList `gorm:"type:text" json:"genres"`
	Cast            StringList `gorm:"type:text" json:"cast"`
	CastDetails     []CastMember `gorm:"serializer:json" json:"cast_details"`
	Director        string     `json:"director"`
	Runtime         int        `json:"runtime"`
	ContentRating   string     `json:"content_rating"`
	TmdbRating      float64    `json:"tmdb_rating"`
	ImdbRating      float64    `json:"imdb_rating"`
	TmdbVotes       int        `json:"tmdb_votes"`
	Metascore       int        `json:"metascore"`
	PosterPath      string     `json:"poster_path"`
	FanartPath      string     `json:"fanart_path"`
	Status          string     `gorm:"default:'unmatched'" json:"status"`
	EpisodeOrdering string     `gorm:"default:'aired'" json:"episode_ordering"`
	TrailerURL      string     `json:"trailer_url"`
	Seasons         []Season   `gorm:"foreignKey:ShowID" json:"seasons"`
}

type Season struct {
	ID           uint   `gorm:"primaryKey;index" json:"id"`
	ShowID       uint   `gorm:"index" json:"show_id"`
	SeasonNumber int    `gorm:"index" json:"season_number"`
	EpisodeCount int    `gorm:"default:0" json:"episode_count"`
	PosterPath   string `json:"poster_path"`
	Episodes     []Episode `gorm:"foreignKey:SeasonID" json:"episodes"`
}

type Episode struct {
	ID               uint   `gorm:"primaryKey;index" json:"id"`
	SeasonID         uint   `gorm:"index" json:"season_id"`
	EpisodeNumber    int    `gorm:"index" json:"episode_number"`
	Title            string `json:"title"`
	Plot             string `json:"plot"`
	AirDate          string `json:"air_date"`
	FilePath         string `gorm:"uniqueIndex" json:"file_path"`
	OriginalFilename string `json:"original_filename"`
	ThumbnailPath    string `json:"thumbnail_path"`
	Resolution       string `json:"resolution"`
	VideoCodec       string `json:"video_codec"`
	AudioCodec       string `json:"audio_codec"`
	Missing          bool `gorm:"default:false" json:"missing"`
}

type BackgroundTask struct {
	ID        string `gorm:"primaryKey;index" json:"id"`
	Name      string `json:"name"`
	Status    string `gorm:"default:'queued'" json:"status"`
	Progress  int    `gorm:"default:0" json:"progress"`
	Total     int    `gorm:"default:100" json:"total"`
	Message   string `json:"message"`
	Duration  float64 `json:"duration"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}
